arangod/Pregel/Actor/MPSCQueue.h

```diff
@@ -0,0 +1,122 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2022 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Markus Pfeiffer
+////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <memory>
+#include <atomic>
+
+// This lock-free multi-producer-single-consumer-queue is inspired by the
+// code published at
+//
+// https://www.1024cores.net/home/lock-free-algorithms/queues/intrusive-mpsc-node-based-queue
+//
+// under a Apache-2.0 license.
```

Comment:
I am not sure I understand your argument. The Vyukov MPSC queue is pretty simple already. Yes, a stack with a reverse might be simpler, but has more contention in push (as you mentioned) and linear overhead for the reverse. This MPSC queue is not linearizable, but that doesn't matter in our case. The way it is used it is not just lock-free, but even wait-free.  It would be quite simple to extend this algorithm for batching if we need it.

With n SPSC queue you not longer have a strict FIFO order.

Reply:
You have allocation for task, and kick (which commonly have internal synchronization and probably allocation in your case) on every push.
So more contention bad argument.

Linear overhead sounds strange, because in general it single additional operation for every task.

About simple, I don't think you right, you known with other more difficult algorithms, but in general pop not simple in Vyukov queue.

And talking about wait-free incorrect, wait-free about guarantees, not about speed.

Anyway, if we talk about this case, it have (incorrect) synchronization on every kick (if (cas)) and probably additional synchronization in push to executor.

Reply:
You always have the allocation for the task, but since the queue itself is intrusive there is not additional allocation there. The scheduler itself is lock-free. ATM it uses the boost lock-free queue which handles its own internal freelist. So yes, pushing tyo the scheduler may require an allocation, but that is negligible. FWIW - I was thinking about using a different queue in the scheduler instead which does not require node allocations (since we only store pointers in the queue).

The reverse is linear in the number of entries in the queue. Granted, I also don't see this as a problem, but it is a difference in complexity to the current algorithm. Yes, the pop is more complex than with an MPSC stack, but not overly so.

I am well aware that lock-freedom/wait-freedom is about progress guarantees and not performance. I never mentioned anything about performance. That said, the current implementation is wait-free population oblivious - the push consists of only two instructions - it doesn't get much faster than this (at least on x86/x64; on ARM the exchange has to be simulated with a LL/SC retry loop).

Reply:
> You always have the allocation for the task (and you will have that one with every other data structure as well).

It's not required, it can be on stack, or on another allocated memory.

> but since the queue itself is intrusive there is no additional allocation there

Agree, but I don't talk about it.

> The scheduler itself is lock-free. ATM it uses the boost lock-free queue which handles its own internal freelist. So yes, pushing tyo the scheduler may require an allocation, but that is negligible. FWIW - I was thinking about using a different queue in the scheduler instead which does not require node allocations (since we only store pointers in the queue).

It depends about what scheduler do we use, in arangodb thread pool it's not very effective:
1. Always make allocation for functor: https://github.com/arangodb/arangodb/blob/devel/arangod/Scheduler/Scheduler.h#L175 doesn't have some free list node or something like that
2. Then it makes push to very slow boost lockfree queue, which use free list node etc
3. Then a lot of atomics loads/writes
4. Then sometime try to wakeup waiting thread (lock mutex etc)

> FWIW - I was thinking about using a different queue in the scheduler instead which does not require node allocations (since we only store pointers in the queue).

It's can be simple intrusive forward list under mutex, it already be a lot of better.
Of course it can be even better with scheduler like in golang/rust tokio for example. But this schedulers doesn't have different priority.
So I think better to start with few simple intrusive forward lists.

> That said, the current implementation is wait-free population oblivious - the push consists of only two instructions - it doesn't get much faster than this (at least on x86/x64; on ARM the exchange has to be simulated with a LL/SC retry loop).

No, from practical side yes, but formally no, Vyukov mentioned it:
> Push function is blocking wrt consumer. I.e. if producer blocked in (*), then consumer is blocked too. Fortunately 'window of inconsistency' is extremely small - producer must be blocked exactly in (*). Actually it's disadvantage only as compared with totally lockfree algorithm. It's still much better lockbased algorithm.

> on ARM the exchange has to be simulated with a LL/SC retry loop

Btw not true for arm starts from v8.1, if I remember correctly

I agree with Vyukov queue better for mpsc scenario in vacuum, because smaller contention in push, but currently I don't see the reason to write more difficult and less effective (in our scenario) code.

Reply:
> It's not required, it can be on stack, or on another allocated memory.
No, it cannot be on the stack if we just want to push the message, schedule the handler and go in with some other work.

Reply:
It can if we know actor in which we push message handle it before end of scope.

But yes for actors, it's should be pretty rare